File: dsa/li16_design/L355_DesignTwitter.java

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;
import java.util.Queue;

public class L355_DesignTwitter {

	public static void main(String[] args) {
		L355_DesignTwitter obj = new L355_DesignTwitter();

		obj.postTweet(1, 5);
		System.out.println(obj.getNewsFeed(1));
		obj.follow(1, 2);
		obj.postTweet(2, 6);
		System.out.println(obj.getNewsFeed(1));
		obj.unfollow(1, 2);
		System.out.println(obj.getNewsFeed(1));
	}

    // Problem: https://leetcode.com/problems/design-twitter/
    // Idea: Use pq to order tweets based on time.
	Map<Integer, HashSet<Integer>> follows;
    Map<Integer, ArrayList<Twit>> twits;
    int time = 0;
    public L355_DesignTwitter() {
        follows = new HashMap<>();
        twits = new HashMap<>();
    }

    public void postTweet(int userId, int tweetId) {
        time++;
        twits.putIfAbsent(userId, new ArrayList<>());
        twits.get(userId).add(new Twit(tweetId, time));
    }

    public List<Integer> getNewsFeed(int userId) {
        List<Integer> tweets = new ArrayList<>();
        Queue<Twit> pq = new PriorityQueue<>((t1,t2)->t2.time-t1.time);

        // user tweets
        List<Twit> ts = twits.getOrDefault(userId, new ArrayList<>());
        for(int i=ts.size()-1; i>=0; i--){
            pq.add(ts.get(i));
        }

        HashSet<Integer> following = follows.getOrDefault(userId, new HashSet<>());
        for(int followeeId: following){
            ts = twits.getOrDefault(followeeId, new ArrayList<>());
            for(int i=ts.size()-1; i>=0; i--){
                pq.add(ts.get(i));
            }
        }

        int count=0;
        while(count<10 && !pq.isEmpty()){
            tweets.add(pq.remove().twittId);
            count++;
        }
        return tweets;
    }

    public void follow(int followerId, int followeeId) {
        follows.putIfAbsent(followerId, new HashSet<>());
        follows.get(followerId).add(followeeId);
    }

    public void unfollow(int followerId, int followeeId) {
        if(follows.containsKey(followerId)){
            follows.get(followerId).remove(followeeId);
        }
    }

}
class Twit{
    int twittId;
    int time;
    Twit(int twittId, int time){
        this.twittId = twittId;
        this.time = time;
    }
}

File: dsa/concepts/16_design/python/l355_design_twitter.py
# Problem: https://leetcode.com/problems/design-twitter/
# Idea:
# - Maintain tweets with timestamps
# - Maintain follow relationships
# - Use a max-heap (priority queue) ordered by time to build news feed
#
# postTweet: O(1)
# follow / unfollow: O(1)
# getNewsFeed: O(n log n) in worst case (acceptable for interview)

import heapq
from collections import defaultdict

class Twitter:

    class Twit:
        def __init__(self, tweetId, time):
            self.tweetId = tweetId
            self.time = time

        # for max heap (reverse time)
        def __lt__(self, other):
            return self.time > other.time

    def __init__(self):
        self.follows = defaultdict(set)      # user -> set of followees
        self.twits = defaultdict(list)       # user -> list of tweets
        self.time = 0

    def postTweet(self, userId: int, tweetId: int) -> None:
        self.time += 1
        self.twits[userId].append(self.Twit(tweetId, self.time))

    def getNewsFeed(self, userId: int):
        pq = []

        # user's own tweets
        for twit in self.twits[userId]:
            heapq.heappush(pq, twit)

        # followees' tweets
        for followee in self.follows[userId]:
            for twit in self.twits[followee]:
                heapq.heappush(pq, twit)

        feed = []
        while pq and len(feed) < 10:
            feed.append(heapq.heappop(pq).tweetId)

        return feed

    def follow(self, followerId: int, followeeId: int) -> None:
        self.follows[followerId].add(followeeId)

    def unfollow(self, followerId: int, followeeId: int) -> None:
        if followerId in self.follows:
            self.follows[followerId].discard(followeeId)


# ---------------- Test (same as Java main) ----------------

obj = Twitter()
obj.postTweet(1, 5)
print(obj.getNewsFeed(1))      # [5]
obj.follow(1, 2)
obj.postTweet(2, 6)
print(obj.getNewsFeed(1))      # [6, 5]
obj.unfollow(1, 2)
print(obj.getNewsFeed(1))      # [5]
